File: myLeetCode/Monotone_stack/503. Next Greater Element II.cpp
// 503. Next Greater Element II
// https://leetcode.com/problems/next-greater-element-ii/

#include <iostream>
#include <vector>
#include <stack>

std::vector<int> nextGreaterElements(std::vector<int>& nums) {
    int n = nums.size();
    std::vector<int> result(n, -1);
    std::stack<int> stk;

    for (int i = 0; i < 2 * n; i++) {
        int num = nums[i % n];
        while (!stk.empty() && num > nums[stk.top()]) {
            int index = stk.top();
            stk.pop();
            result[index] = num;
        }
        if (i < n) {
            stk.push(i);
        }
    }

    return result;
}

int main() {
    std::vector<std::vector<int>> testCases = {
        {1, 2, 1},
        {1, 2, 3, 4, 3},
        {5, 4, 3, 2, 1}
    };

    for (int i = 0; i < testCases.size(); i++) {
        std::vector<int> nums = testCases[i];

        std::vector<int> result = nextGreaterElements(nums);

        std::cout << "Next greater elements: ";
        for (int j = 0; j < result.size(); j++) {
            std::cout << result[j] << " ";
        }
        std::cout << std::endl;
    }

    return 0;
}

File: C++/代码随想录/Monotone_stack/503. Next Greater Element II.py
def nextGreaterElements(nums):
    n = len(nums)
    result = [-1] * n
    stack = []

    for i in range(2 * n):
        while stack and nums[stack[-1]] < nums[i % n]:
            result[stack.pop()] = nums[i % n]
        if i < n:
            stack.append(i)

    return result

# 测试用例
print(nextGreaterElements([1, 2, 1]))  # 应输出 [2, -1, 2]
print(nextGreaterElements([1, 2, 3, 4, 3]))  # 应输出 [2, 3, 4, -1, 4]
